bcipy/signal/evaluate/artifact.py

```diff
@@ -0,0 +1,595 @@
+
+        self.session_triggers = None
+        if session_triggers:
+            self.triggers = session_triggers
+            self.trigger_description = session_triggers[0]
+            self.trigger_time = session_triggers[1]
+            self.trigger_label = session_triggers[2]
+
+            self.session_triggers = mne.Annotations(
+                self.trigger_time, [self.trial_duration] * len(self.trigger_time), self.trigger_description)
+
+        assert len(device_spec.channel_specs) > 0, 'DeviceSpec used must have channels. None found.'
+        self.units = device_spec.channel_specs[0].units
+        assert self.units in self.supported_units, \
+            f'Data loaded in units that cannot be processed. Support units={self.supported_units}'
+        # MNE assumes that data is recorded in Volts. However, many devices use microvolts.
+        # If this value is False, a conversion to volts will be required.
+        self.volts = True if self.units == 'volts' else False
+        self.mne_data = self.raw_data_to_mne(raw_data, volts=self.volts)
+        self.detect_voltage = detect_voltage
+        self.detect_eog = detect_eog
+        self.semi_automatic = semi_automatic
+
+        self.save_path = save_path
+
+    def detect_artifacts(self) -> Tuple[str, float]:
+        """Detect artifacts in the raw data."""
+
+        labels = self.label_artifacts(extra_labels=self.session_triggers)
+        percent_dropped = 0
+
+        # calculate the impact of artifacts on the session triggers by creating epochs around the triggers
+        # and dropping artifacts using MNE
+        if self.session_triggers:
+            log.info('Calculating the impact of artifacts on session triggers.')
+            epochs = mne_epochs(
+                self.mne_data,
+                self.trial_duration,
+                reject_by_annotation=True,
+                preload=True)
+
+            # calculate the percentage of dropped epochs
+            kept_epochs = len(epochs)
+            total_epochs = len(self.trigger_label)
+            percent_dropped = (1 - (kept_epochs / total_epochs)) * 100
+            self.dropped = (
+                f'{kept_epochs}/{total_epochs} retained. {percent_dropped}% epochs dropped due to artifacts.'
+            )
+
+        self.analysis_done = True
+        if self.save_path:
+            self.save_artifacts(overwrite=True)
+            log.info(f'Artifact labelled data saved to {self.save_path}')
+
+        labels = f'{len(labels)} artifacts found in the data.'
+
+        return labels, percent_dropped
+
+    def label_artifacts(
+            self,
+            detect_voltage: bool = True,
+            detect_eog: bool = True,
+            extra_labels: Optional[Annotations] = None) -> Annotations:
+        """Label the artifacts in the raw data."""
+        # Create an empty annotations object to store all the annotations
+        annotations = mne.Annotations(0, 0, 'start')
```

Comment:
This class is very large and it might make sense to split up its responsibilities. One option is to introduce the concept of an `ArtifactLabeler` with a subclass for each type (ex. `VoltageLabeler`, `EOGEventLabeler`). Then this method could iterate through the `ArtifactTypes` of interest, matching each one to a labeler, and combine the results.